### PR TITLE
support ROSAT PSPC (and similar) RMF files with AstroPy 

### PR DESCRIPTION
**Pull request by DougBurke**
# Release Notes

Add explicit tests for reading in, and using, ROSAT PSPC PHA and RMF files. Fix a bug in the AstroPy back-end handling of the ROSAT RMF file.

# Other notes

Hidden away now in this PR is a commit that adds support for a Travis-CI build that uses the CIAO 4.9 Crates module (Python 2.7 only), to provide immediate feedback for changes that might affect the Crates backend. There is no support for XSPEC or ChIPS in this build. This was used to verify that the changes didn't break anything obvious. This commit has been backed out for now, but could be resurrected at a future date (it moved the logic for Travis testing into external scripts rather than all being encoded in the `.travis.yml` file, which could in itself be a useful update even if the CIAO-specific options are not supported).

# Details

The logic for creating the test environment has been moved out of the Travis yaml file and into scripts in the `scripts/` directory. I like the idea, as it potentially makes it easier to re-create a particular Travis test (e.g. if one of them is failing) locally. The CIAO build is fragile - in the sense that it relies on CIAO 4.9 which requires kludges to support working around some installation issues. It could be separated from the bug fix, but was useful to help me make sure I did not break anything on the Crates side (in particular in setting up useful tests).

This PR can be thought of as an follow-up to #358, in that the code it corrects the code added in that commit to handle the more-general case (i.e. the Swift RMF format is "simpler" than the ROSAT PSPC one).

### CODE REVIEW

**Comment by DougBurke** [comment]
This came out of @olaurino code review of #399 - in particular my comment at https://github.com/sherpa/sherpa/pull/399#discussion_r154169174

I felt it worth exploring the rate behavior, and it's probably easier to do so in a separate PR at the moment.

**Comment by DougBurke** [comment]
So, turns out that crates is throwing the error when reading the RMF file. I'll report that internally.

**Comment by DougBurke** [comment]
Well, that was a spectacular success. Perhaps the build logic should be factored out into scripts so that they can be tested separately, and make .travis.yml cleaner?

**Comment by DougBurke** [comment]
About to rebase to deal with change in `sherpa-test-data`. Original tip was cb493a3

**Comment by DougBurke** [comment]
About to rebase to make it look like I knew what I was doing all along. Tip was 88665e4

**Comment by DougBurke** [comment]
About to rebase to simplify the commits. Original tip was 6634dfd

**Comment by DougBurke** [comment]
About to rebase. Original tip was a572d55

**Comment by DougBurke** [comment]
@olaurino - can I remove the pr:hold label from this, or are there still build conflicts?

**Comment by olaurino** [comment]
I don't really know. Remove it and if something comes up we'll address that.

**Comment by olaurino** [comment]
Yeah, unfortunately it must be conflicting with something else, so it doesn't get merged into the snapshot branch. Most likely the issue is with PR 419 since the conflict is in the travis configuration file. Which reminds me I owe you a reply in that thread.

**Comment by DougBurke** [comment]
About to rebase. Original tip was 613a261

**Comment by olaurino** [comment]
@DougBurke I'd be in favor of removing the new travis job and taking the fix. As for moving some of the logic out of the .travis.yml I never thought the setup part was complex enough to do that, but I'd be in favor of that as well.

**Comment by DougBurke** [comment]
@olaurino can you cherry pick 661cd4b into a new PR. I'd split this into two commits to make it easy to do so, but don't have much time this week for Sherpa

**Comment by olaurino** [comment]
It seems to make more sense to reset this to the previous commit. As long as you have a branch pointing to the current tip git shouldn't GC the commit if you want to create a new PR.

**Comment by olaurino** [comment]
Rebasing (was 63d8bce). There, now even GitHub will keep it around forever ;)

**Comment by DougBurke** [comment]
@olaurino - I'm a bit confused here. I've just edited the summary to say "this fixes a bug, but hides away the travis changes", but it seems that now the only things left are the travis changes. What happened to the actual bug fix? Has that already been merged?

**Comment by olaurino** [comment]
Yeah, I think I saved the wrong commit last night, I was just fixing it.

**Comment by olaurino** [comment]
Interesting. It looks like after merging #383 this PR now gets errors because the new test files don't validate.

**Comment by olaurino** [comment]
@DougBurke how does it look?

File: sherpa/astro/data.py [hunk 1]
```diff
@@ -403,6 +403,19 @@ def __setstate__(self, state):
             self.header = None
         self.__dict__.update(state)
 
+    def _validate(self, name, energy_lo, energy_hi, ethresh):
+        """
+        Default validation: do nothing and just pass the arrays through.
```
Comment (olaurino) [inline sherpa/astro/data.py]: Oops, I left this comment by mistake. Waiting for more comments before fixing.

Comment (DougBurke) [inline sherpa/astro/data.py]: I don't think this comment is right. It looks to me like the default is to actually do the validation. Or am I reading this wrong?

File: sherpa/astro/data.py [hunk 2]
```diff
@@ -454,6 +467,16 @@ def get_ylabel(self):
         return 'Counts'
 
 
+# FIXME There are places in the code that explicitly check if an object is an instance of sherpa.astro.data.DataRMF.
+# So it's safer to make DataRosatRMF a subclass of the default class, although in principle they should be siblings
+# and subclasses of the same superclass.
+class DataRosatRMF(DataRMF):
```
Comment (DougBurke) [inline sherpa/astro/data.py]: In principle this is `DataNoValidationRMF`, which is subclassed to `DataRosatRMF`, but I think that's over-engineering for now (the thought being that there could be other cases where we decide to do no validation).

File: sherpa/astro/data.py [hunk 3]
```diff
@@ -403,6 +403,19 @@ def __setstate__(self, state):
...
+        """
+        Default validation: do nothing and just pass the arrays through.
+        Subclasses may override this method to perform different validations.
+
+        :param name: The name/label of the current file
+        :param energy_lo: The lower bounds of the energy bins
+        :param energy_hi: The upper bounds of the energy bins
+        :param ethresh: The lowest energy value, if any
+        :return: A tuple of energy bounds arrays, validated and possibly adjusted
+        """
```
Comment (DougBurke) [inline sherpa/astro/data.py]: We've been using something like the following (although I am never sure what textual label to use for NumPy array here):

    Parameters
    ----------
    name : str
        The name/label of the current file.
    energ_lo, energ_hi : NumPy array
        The lower and upper bounds of the energy bins. The arrays must have
         the same size
    ethresh : float
        The lowest valid energy value.

    Returns
    -------
    energ_lo, energ_hi : NumPy array
        The energy values to use for the bin boundaries

Comment (olaurino) [reply sherpa/astro/data.py]: Right. I thought I had already configured my IDE to use the right style. I'll check again.

**Comment by DougBurke** [comment]
The only quibbles I have are over comments and names, but neither are serious.

File: sherpa/astro/data.py (hunk 2 above)
Comment (olaurino) [reply sherpa/astro/data.py]: In principle, I agree.

**Comment by olaurino** [comment]
Pushed a new commit fixing the docstring. For the class hierarchy I agree, but we don't need the intermediate class at this point. If we do in the future it will be easy to introduce one without changing the API and without affecting existing code, I think.

**Comment by olaurino** [comment]
@DougBurke the special test that checked for the crates exception is now failing as the upstream bug was fixed. I am going to remove the test, but I wanted to double check we don't need to make simplifications to the code somewhere else. I don't think so, as the tests were added with this same PR, but wanted to double check.

**Comment by olaurino** [comment]
I forgot, there was a test being skipped as it didn't work on crates, I am also going to remove the skip condition.

**Pull request merged**